examples/martingale.py

```diff
@@ -0,0 +1,183 @@
+        self.streak_increasing = True
+
+        # When this variable is not None, we have an order open
+        self.current_order = None
+
+        # The closing price of the last aggregate we saw
+        self.last_price = 0
+
+        # The connection to the Alpaca API
+        self.api = tradeapi.REST(
+            self.key_id,
+            self.secret_key,
+            self.base_url
+        )
+
+        # Cancel any open orders so they don't interfere with this script
+        self.api.cancel_all_orders()
+
+        try:
+            self.position = int(self.api.get_position(self.symbol).qty)
+        except:
+            # No position exists
+            self.position = 0
+
+        # Figure out how much money we have to work with, accounting for margin
+        account_info = self.api.get_account()
+        self.equity = float(account_info.equity)
+        self.margin_multiplier = float(account_info.multiplier)
+        total_buying_power = self.margin_multiplier * self.equity
+        print(f'Initial total buying power = {total_buying_power}')
+
+
+    def start_trading(self):
+        conn = tradeapi.StreamConn(
+            self.key_id,
+            self.secret_key,
+            self.base_url
+        )
+
+        # Listen for second aggregates and perform trading logic
+        @conn.on(r'A$', [self.symbol])
+        async def handle_agg(conn, channel, data):
+            self.tick_index += 1
+            if self.tick_index == self.tick_size - 1:
+                # It's time to update
+                self.tick_index = 0
+
+                # Update price info
+                tick_open = self.last_price
+                tick_close = data.close
+                self.last_price = tick_close
+
+                # Update streak info
+                diff = truncate(tick_close, 2) - truncate(tick_open, 2)
+                if diff != 0:
+                    # There was a meaningful change in the price
+                    self.streak_count += 1
+                    increasing = tick_open > tick_close
+                    if self.streak_increasing != increasing:
+                        # It moved in the opposite direction of the streak.
+                        # Therefore, the streak is over, and we should reset.
+
+                        # Empty out the position
+                        self.send_order(0)
+
+                        # Reset variables
+                        self.streak_increasing = increasing
+                        self.streak_start = tick_open
+                        self.streak_count = 0
+                    else:
+                        # Calculate the number of shares we want to be holding
+                        total_buying_power = self.equity * \
+                            self.margin_multiplier
+                        target_value = (2**self.streak_count) * \
+                            (self.base_bet / 100) * total_buying_power
+                        if target_value > total_buying_power:
```

Comment:
Without removing one share from the final quantity, sometimes it was trying to order one more share than it could. I think it might have had to do with a discrepancy between this script's view of the price and Alpaca's view. Either way, I'd rather just set the max quantity it can order as one less than what it thinks is possible than dig too much into it.